src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -229,119 +277,187 @@ def _backchain_one_pass(
-        must make its add effects more specific. Note that we are
-        assuming all variables in the parameters of the PNAD will appear
-        in either the option arguments or the add effects. This is in
-        contrast to strips_learning.py, where delete effect variables
-        also contribute to parameters. If
-        ground_eff_subset_necessary_eff is True, we want to find a
-        grounding that achieves some subset of the
-        necessary_add_effects. Else, we want to find a grounding that is
-        some superset of the necessary_add_effects and also such that
-        the ground operator's add effects are always true in the
-        segment's final atoms.
+    ) -> Tuple[Optional[PartialNSRTAndDatastore],
+               Optional[_GroundSTRIPSOperator]]:
+        """Find a mapping from the variables in one of the PNAD's add effects
+        and option to the objects in necessary_add_effects and the segment's
+        option.
+
+        If a mapping exists, we don't need to modify this PNAD.
+        Otherwise, we must make its add effects more specific. Note that
+        we are assuming all variables in the parameters of the PNAD will
+        appear in either the option arguments or the add effects. This
+        is in contrast to strips_learning.py, where delete effect
+        variables also contribute to parameters. If
+        ground_eff_subset_necessary_eff is True, we want to find the
+        best grounding that achieves some subset of the
+        necessary_add_effects. Else, we want to find the best grounding
+        that is some superset of the necessary_add_effects and also such
+        that the ground operator's add effects are always true in the
+        segment's final atoms. Note that we score the groundings and
+        pick the best one since there may be multiple operators that fit
+        these necessary add effects.
         """
         objects = list(segment.states[0])
+
         option_objs = segment.get_option().objects
-        isub = dict(zip(pnad.option_spec[1], option_objs))
-        # Loop over all groundings.
-        for ground_op in utils.all_ground_operators_given_partial(
-                pnad.op, objects, isub):
-            if not ground_op.preconditions.issubset(segment.init_atoms):
-                continue
-            if ground_eff_subset_necessary_eff:
-                if not ground_op.add_effects.issubset(necessary_add_effects):
+        # Loop over all ground operators, looking for the most
+        # rational match for this segment.
+        best_score = float("inf")
+        best_pnad = None
+        best_ground_pnad = None
+        for pnad in pnads:
+            param_opt, opt_vars = pnad.option_spec
+            assert param_opt == segment.get_option().parent
+            isub = dict(zip(opt_vars, option_objs))
+            # Handle keep effect substitution correctly
+            # NOTE: code copied directly from _recompute_datastores.
```

Comment:
yeah okay, we should definitely merge this function with recompute_datastores before merging

Reply:
will do in new PR!